test/unit/test_keyring_raw_aes.py

```diff
@@ -0,0 +1,343 @@
+            key_namespace=_PROVIDER_ID,
+            key_name=_KEY_ID,
+            wrapping_algorithm=WrappingAlgorithm.AES_256_GCM_IV12_TAG16_NO_PADDING,
+            wrapping_key=WrappingAlgorithm.AES_256_GCM_IV12_TAG16_NO_PADDING,
+        )
+    assert exc_info.errisinstance(TypeError)
+
+
+def test_on_encrypt_when_data_encryption_key_given(patch_generate_data_key):
+    test_raw_aes_keyring = RawAESKeyring(
+        key_namespace=_PROVIDER_ID,
+        key_name=_KEY_ID,
+        wrapping_algorithm=WrappingAlgorithm.AES_256_GCM_IV12_TAG16_NO_PADDING,
+        wrapping_key=_WRAPPING_KEY,
+    )
+
+    test_raw_aes_keyring.on_encrypt(encryption_materials=get_encryption_materials_with_data_encryption_key())
+    # Check if keyring is generated
+    assert not patch_generate_data_key.called
+
+
+def test_keyring_trace_on_encrypt_when_data_encryption_key_given():
+    test_raw_aes_keyring = RawAESKeyring(
+        key_namespace=_PROVIDER_ID,
+        key_name=_KEY_ID,
+        wrapping_algorithm=WrappingAlgorithm.AES_256_GCM_IV12_TAG16_NO_PADDING,
+        wrapping_key=_WRAPPING_KEY,
+    )
+
+    test = test_raw_aes_keyring.on_encrypt(encryption_materials=get_encryption_materials_with_data_encryption_key())
+
+    for keyring_trace in test.keyring_trace:
+        if keyring_trace.wrapping_key.key_info == _KEY_ID:
+            # Check keyring trace does not contain KeyringTraceFlag.WRAPPING_KEY_GENERATED_DATA_KEY
+            assert KeyringTraceFlag.WRAPPING_KEY_GENERATED_DATA_KEY not in keyring_trace.flags
+
+
+def test_on_encrypt_when_data_encryption_key_not_given():
+
+    test_raw_aes_keyring = RawAESKeyring(
+        key_namespace=_PROVIDER_ID,
+        key_name=_KEY_ID,
+        wrapping_algorithm=WrappingAlgorithm.AES_256_GCM_IV12_TAG16_NO_PADDING,
+        wrapping_key=_WRAPPING_KEY,
+    )
+
+    original_number_of_encrypted_data_keys = len(
+        get_encryption_materials_without_data_encryption_key().encrypted_data_keys
+    )
+
+    test = test_raw_aes_keyring.on_encrypt(encryption_materials=get_encryption_materials_without_data_encryption_key())
+
+    # Check if data key is generated
+    assert test.data_encryption_key and test.data_encryption_key is not None
+
+    generated_flag_count = 0
+    encrypted_flag_count = 0
+
+    for keyring_trace in test.keyring_trace:
+        if (
+            keyring_trace.wrapping_key.key_info == _KEY_ID
+            and KeyringTraceFlag.WRAPPING_KEY_GENERATED_DATA_KEY in keyring_trace.flags
+        ):
+            # Check keyring trace contains KeyringTraceFlag.WRAPPING_KEY_GENERATED_DATA_KEY
+            generated_flag_count += 1
+        if KeyringTraceFlag.WRAPPING_KEY_ENCRYPTED_DATA_KEY in keyring_trace.flags:
+            encrypted_flag_count += 1
+
+    assert generated_flag_count == 1
+
+    assert len(test.encrypted_data_keys) == original_number_of_encrypted_data_keys + 1
+
+    assert encrypted_flag_count == 1
+
+
+def test_on_decrypt_when_data_key_given(patch_decrypt_data_key):
```

Comment:
Let's parameterize this for an additional decryption materials case:
- [x] data key in input materials and EDKs provided
- [ ] data key in input materials and *no* EDKs provided